### PR TITLE
Fix performance regression of buf_filled_with function

### PR DESCRIPTION
**Pull request by shizhx**
because of previous modification #204  of `strings.py`, performance of `buf_filled_with` is bad.
but anyway, `mmap object` has no `count` method, I have to implement it with pure python.
now, `mmap object` + `buf_filled_with` is even **faster** than `read` + `count`.

here is test script and full performance testing report: https://github.com/shizhx/others/tree/master/flare-floss_test

part of performance testing report:
use `read` with `count`
-------------------------------
```
2048MB sample:
         1918 function calls (1898 primitive calls) in 11.757 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.051    0.051   11.757   11.757 strings_count.py:1(<module>)
        1    0.000    0.000   11.702   11.702 strings_count.py:89(main)
        1    7.188    7.188    7.188    7.188 {method 'read' of 'file' objects}
        2    4.514    2.257    4.514    2.257 {method 'count' of 'str' objects}
        1    0.000    0.000    2.297    2.297 strings_count.py:33(extract_ascii_strings)
        1    0.000    0.000    2.217    2.217 strings_count.py:60(extract_unicode_strings)
```

use `mmap` with `buf_filled_with`
---------------------------------------------
```
2048MB sample:
         3147648 function calls (3147628 primitive calls) in 8.093 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.043    0.043    8.093    8.093 strings_buf_filled_with.py:1(<module>)
        1    0.000    0.000    8.047    8.047 strings_buf_filled_with.py:90(main)
        2    2.636    1.318    8.046    4.023 strings_buf_filled_with.py:15(buf_filled_with)
  1048576    5.271    0.000    5.271    0.000 {method 'count' of 'str' objects}
        1    0.001    0.001    5.191    5.191 strings_buf_filled_with.py:34(extract_ascii_strings)
        1    0.000    0.000    2.856    2.856 strings_buf_filled_with.py:61(extract_unicode_strings)
2097660/2097656    0.140    0.000    0.140    0.000 {len}
```

**please perform code review carefully and help me to improve these**

relevant issues: #205 #222

### CODE REVIEW

**Comment by pilate** [comment]
We can pick up a little more performance by caching the last chunk and doing a straight comparison. 

This looks like it's actually several times faster than count for both strings and mmap objects:

```python
def buf_filled_with(buf, c):
    prev_chunk = ""
    for offset in xrange(0, len(buf), SLICE_SIZE):
        new_chunk = buf[offset: offset + SLICE_SIZE]
        new_len = len(new_chunk)

        if prev_chunk and (prev_chunk[:new_len] == new_chunk):
            continue

        if not new_chunk.count(c) == new_len:
            return False

        prev_chunk = new_chunk

    return True
```

Need to validate I'm not overlooking something simple, though.

File: floss/strings.py [hunk 1]
```diff
@@ -7,16 +7,29 @@
...
 
 def buf_filled_with(buf, c):
-    for i in range(len(buf)):
-        if buf[i] != c:
+    if hasattr(buf, 'count'):
+        return buf.count(c) == len(buf)
+
+    begin = 0
+    end = 0
+    while begin < len(buf):
```
Comment (williballenthin) [inline floss/strings.py]: `for end in xrange(0, len(buf), SLICE_SIZE):`

**Comment by williballenthin** [comment]
@shizhx thank you very much for taking ownership of the testing and enhancement of this routine. i think your patch is moving in the right direction. the attr-checking for `count` is just what i had in mind when i made a similar suggestion in the issue. processing the data in chunks is a neat idea.

i'm interested in the results that @pilate reports. i like that his implementation is short, and following the same lines of thinking in using chunked data.

aside from further testing, my only thought is: can we avoid the creation of temporary str/arrays? seems each time we take a chunk, it has to be copied. avoiding the copy *could* improve performance, but it might also be too small to matter.

my first thought was using a `memoryview`, but apparently these aren't supported for `mmap` objects (http://bugs.python.org/issue9229).

second thought is using `buf.index(dupe_chunk, start, end)` and catching the `ValueError` if there's no match (immediately return false, no loopy exceptions).

**Comment by williballenthin** [comment]
travis build failure is not related to this patch.

**Comment by shizhx** [comment]
@pilate  great code! It is really faster than  "count implementation". I think it is the best implementation at the moment.

@williballenthin completely reasonable thinking.

I am not very similar with Python, I will create an issue for this kind of problem next time.

**Comment by pilate** [comment]
@williballenthin I tried to avoid temporary strings using mmap.find, but it was far slower than any of the existing solutions.

**Comment by williballenthin** [comment]
closed in #226

**Pull request closed without merging**